connectors/cli.py

```diff
@@ -102,13 +102,14 @@ async def _start_service(config, args, loop):
     services = [SyncService(config, args)]
     for sig in (signal.SIGINT, signal.SIGTERM):
 
-        def _shutdown(sig_name):
+        async def _shutdown(sig_name):
             logger.info(f"Caught {sig_name}. Graceful shutdown.")
             for service in services:
                 logger.info(f"Shutdown {service.__class__.__name__}...")
-                service.stop()
+                await service.stop()
+            return
 
-        loop.add_signal_handler(sig, functools.partial(_shutdown, sig.name))
+        loop.add_signal_handler(sig, lambda: asyncio.ensure_future(_shutdown(sig.name)))
```

Comment:
why do you need to do this?

Reply:
`_shutdown` is made async in this change, thus it's not usable by `add_signal_handler` as it was before.

I made it work with the current code, but not sure if it's the best way to do so (if you revert this change you'll see tests break).

Reply:
we can't start a future in this signal handler it needs to block (and you are not waiting for it so that's a problem)
let's sync about this

Reply:
I took this code from StackOverflow: https://stackoverflow.com/questions/23313720/asyncio-how-can-coroutines-be-used-in-signal-handlers

I don't have enough understanding yet about it, of course

Reply:
Do this instead:

```
tasks = []
for sig in (signal.SIGINT, signal.SIGTERM):
    def _shutdown(sig_name):
        for service in services:
            tasks.append(asyncio.create_task(service.stop()))
    loop.add_signal_handler(sig, functools.partial(_shutdown(sig.name)))

for service in services:
    tasks.append(_run_service(service))  

await asyncio.gather(*tasks)
```

Reply:
Will it be easier if we assign this to `self.sync_job_runner`

https://github.com/elastic/connectors-python/blob/3e5c01202539bc8a3eac79f74d1b107242f9ae24/connectors/services/sync.py#L86

And in `stop()` method of `SyncService`, we add
```python
    if self.sync_job_runner is not None:
        self.sync_job_runner.suspend()
```

Reply:
@tarekziade I'm not sure I understand how the approach you described works - `await asyncio.gather(*tasks)` intuitively should not wait on new tasks added to the `tasks`. I tried to make a small script to try it and it does not await for newly added tasks: https://gist.github.com/artem-shelkovnikov/e949b51657377b74a6dbce03646ca911

I've seen a diferent approach [here](https://stackoverflow.com/questions/61208300/asyncio-wait-for-each-task-in-a-dynamic-list-to-finish-or-be-cancelled) with gathering tasks as long as any are present, but it's pretty complex.

I also tested the suspension logic how it's done now and added `await asyncio.sleep(10)` in it inside `_sync_suspended` and verified that it's actually awaited. Is there even a problem to fix here?

Reply:
@wangch079 it can be encapsulated like that, but the `suspend` function is async and `loop. add_signal_handler` is not supposed to work with async function, if I got Tarek's concern correctly.

Reply:
synced with @artem-shelkovnikov -- we'll adapt `ConcurrentTasks` and use it